python/pdo/submitter/sawtooth/pdo_protos/protobufs/pdo_contract_ccl.proto

```diff
@@ -88,4 +88,7 @@ message CCL_TransactionPayload {
 
     // PDO signature to be verified with contract creator key
     string pdo_signature = 6;
+
+    // Only valid for initialize
+    string metadata_hash = 7;
```

Comment:
Given that the metadata is static, wouldn't it be more suitable to put the `metadata_hash` in `pdo_contract_registry.proto` (hence, in a structure used at registration time)?

There are some issues with that, but arguably due to the registration procedure.
1. the `contract_id` is not available. However, now the contract has a signing key, which should be unique. So, is the contract id needed?
2. the contract is not invoked (or better, the contract keys are not generated) at registration time. Only the contract enclave is invoked to sign the provisioning. But this can be easily changed.
3. when using multiple enclaves, the contract key generation should return the same key. This is not currently the case, because a random key is generated. However, it _could_ be the case, considering that the provisioning services provide a source of randomness, from which we already generate the state encryption key.

Reply:
The problem is that you cannot create the metadata at contract registration. Remember that call happens before any state encryption keys are even created.

Reply:
I should clarify point 2.
You're right about the registration.
I wanted actually to refer to the `add enclave` step. In that step the state encryption key is put together, signed and encrypted by the enclave.

Reply:
again... i think there is no way to collect all the information together until the initialize method is called on the contract. creating the state encryption key is NOT the same operation as creating the keys for secure communication and verifiable identity of the contract objects. With your approach we would need to use an MPC to generate those (since they have to be stored in state which doesn't actually exist at that point).

Reply:
and regarding 3... deterministic generation of the keys would be possible, but I don't see how that would improve the security unless we were concerned about generating enough keys (by the owner of the object) to manipulate the final value. i think we have agreed (given that we have similar issues in other places) that the computation is sufficiently complex that it would take months to compute. And, the best solution would be to throttle operations at the enclave.